### PR TITLE
Scan url with QR scanner

### PR DESCRIPTION
**Pull request by krisc**
This PR is for this bounty https://github.com/status-im/status-react/issues/5776

It fixes the issue as described, but I left it WIP because I'm not sure if additional checks have to be made or not. For example, if the qr-code returns data `30032621` should this show an error message or should it just navigate to whatever that data is in the browser? 

If it should show an error, do we have a function already in status that checks for valid urls?

### CODE REVIEW

**Comment by mandrigin** [comment]
@krisc thanks for your contribution! maybe @yenda or @flexsurfer can help you with the error function. But I would check if what is returned is similar to a URL, and show error otherwise.

File: env/dev/env/config.cljs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 (ns env.config)
 
 (def figwheel-urls {:android "ws://localhost:3449/figwheel-ws",
- :ios "ws://192.168.0.9:3449/figwheel-ws",
+ :ios "ws://192.168.1.84:3449/figwheel-ws",
```
Comment (mandrigin) [inline env/dev/env/config.cljs]: this change should be reverted

**Review by mandrigin — CHANGES_REQUESTED** [review]

**Comment by mandrigin** [comment]
@krisc also, the formatting should be fixed (that's why CI builds are failing). I recommend running `lein cljfmt fix` to fix formatting.

**Comment by krisc** [comment]
> @krisc also, the formatting should be fixed (that's why CI builds are failing). I recommend running `lein cljfmt fix` to fix formatting.

Oh, thanks! wasn't aware of that command. I get that error every time and didn't know why.

**Comment by krisc** [comment]
> @krisc thanks for your contribution! maybe @yenda or @flexsurfer can help you with the error function. But I would check if what is returned is similar to a URL, and show error otherwise.

@yenda or @flexsurfer , is there a function somewhere in status that checks for vaild URLs? Perhaps someone could direct me to the code where URLs in chat messages are rendered as blue and underlined?

For now, I'm just checking if the string has a "." in it.

**Comment by krisc** [comment]
@mandrigin This should be it unless I should be error checking a different way. And FYI, the case where browser selection is canceled and then being able to scan another qr-code depends on code that came from this PR https://github.com/status-im/status-react/pull/7745

This PR pretty much has the aforementioned PR merged since it shares so much code.  The other PR should probably be merged to upstream first.

**Comment by mandrigin** [comment]
@krisc sounds good, then let's first land #7745 and then this one. I will take a look at the code today.

File: src/status_im/ui/screens/add_new/models.cljs [hunk 2]
```diff
@@ -4,19 +4,16 @@
...
-        {:utils/show-popup {:title      (i18n/label :t/unable-to-read-this-code)
-                            :content    (i18n/label :t/use-valid-qr-code {:data data})
-                            :on-dismiss #(re-frame/dispatch [:navigate-to-clean :home])}})))
-
-(fx/defn handle-qr-code
-  [cofx data]
-  (fx/merge cofx
-            (navigation/navigate-to-clean :home {})
-            (process-qr-code data)))
+        (if (clojure.string/includes? data ".")
```
Comment (mandrigin) [inline src/status_im/ui/screens/add_new/models.cljs]: I think you can use something like that to check for URLs:
```clojure
(defn is-url [url] 
  (try 
    (and (js/URL. url) true) 
  (catch js/Error e false)))
```

**Review by mandrigin — CHANGES_REQUESTED** [review]
the condition needs to be changed, I think, looks good otherwise!

File: src/status_im/ui/screens/add_new/models.cljs (hunk 2 above)
Comment (krisc) [reply src/status_im/ui/screens/add_new/models.cljs]: hmm this seems to always return true

Comment (mandrigin) [reply src/status_im/ui/screens/add_new/models.cljs]: hmm, I tried it in a repl and for me it returns `false` for, say, “aaa” and `true` for “http://google.com"...

Comment (krisc) [reply src/status_im/ui/screens/add_new/models.cljs]: hmm I'm using figwheel for status-im, and returns true for anything. I did try in a web repl, and it does return `false` for something like "aaa". which repl did you use?

Comment (krisc) [reply src/status_im/ui/screens/add_new/models.cljs]: I think this should be handled the same way a URL is handled when tapped on by a user in chat so I would like to see where in the code that renders a chat message as a link. for example, if a message contains "aaaa google.com" then the 'google.com' becomes underlined. I'm having trouble trying to find where this happens :/

Comment (mandrigin) [reply src/status_im/ui/screens/add_new/models.cljs]: @yenda maybe you know?

**Comment by krisc** [comment]
@mandrigin @yenda I ended up using `extensions/valid-uri?` but was not yet able to test it because I'm having issues with `make shell`

**Review by mandrigin — APPROVED** [review]
I think that's good enough for now.

**Comment by krisc** [comment]
Derp, I should have read the code https://github.com/status-im/status-react/blob/64b63d5593cebd7382b6894a71848589f9628e67/src/status_im/extensions/core.cljs#L865

It will always return `false` unless it's a status extension.

> I think this should be handled the same way a URL is handled when tapped on by a user in chat so I would like to see where in the code that renders a chat message as a link. for example, if a message contains "aaaa google.com" then the 'google.com' becomes underlined. I'm having trouble trying to find where this happens :/

If someone could point me to where this happens, that would be great!

File: src/status_im/ui/screens/add_new/models.cljs (hunk 2 above)
Comment (yenda) [reply src/status_im/ui/screens/add_new/models.cljs]: we have a `regx-url` in `constants.cljs` to check for urls

**Comment by yenda** [comment]
@krisc this is done in `status-im.chat.models.message-content`, the message string is parsed and urls are replaced by `[:link "the url"]`, to find urls the regexp used is in `constants/regex-url`

**Comment by krisc** [comment]
> @krisc this is done in `status-im.chat.models.message-content`, the message string is parsed and urls are replaced by `[:link "the url"]`, to find urls the regexp used is in `constants/regex-url`

Exactly what I was looking for. Thanks!

**Comment by krisc** [comment]
Ok, this should be it!

File: src/status_im/chat/models/message_content.cljs [hunk 3]
```diff
@@ -87,12 +88,15 @@
...
   [{:keys [text] :as content}]
+  ; (log/debug "enrich-content text = " text)
+  ; (log/debug "enrich-content content = " content)
   (let [[_ metadata] (reduce (fn [[text metadata] [type regex]]
                                (if-let [matches (query-regex regex text)]
                                  [(clear-ranges matches text) (assoc metadata type matches)]
                                  [text metadata]))
                              [text {}]
                              (into stylings actions))]
+    ; (log/debug "enrich-content metadata = " metadata)
```
Comment (mandrigin) [inline src/status_im/chat/models/message_content.cljs]: remove commented-out code

File: src/status_im/chat/models/message_content.cljs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 (ns status-im.chat.models.message-content
   (:require [clojure.string :as string]
+            [taoensso.timbre :as log]
```
Comment (mandrigin) [inline src/status_im/chat/models/message_content.cljs]: unused import

**Review by mandrigin — CHANGES_REQUESTED** [review]
a few tiny changes, looks good otherwise!

File: src/status_im/chat/models/message_content.cljs (hunk 3 above)
Comment (krisc) [reply src/status_im/chat/models/message_content.cljs]: oh yeah forgot to remove my debug messages

**Review by mandrigin — APPROVED** [review]
looks good now! just don't forget to squash the commits ;-)

**Comment by krisc** [comment]
@mandrigin not sure why it's saying "jenkins/prs/android-e2e — This commit cannot be built". I ran `lein cljfmt fix` before committing

**Comment by mandrigin** [comment]
@krisc looks like a random thing, restarted it

**Comment by krisc** [comment]
@mandrigin BTW is there a way I can get access to the jenkins build logs? When I click on "Details" is says `
Access Denied
krisc is missing the Overall/Read permission`

**Comment by mandrigin** [comment]
@krisc I think the access is limited until we audit our build logs for sensitive info... We will get there, it isn't just done yet.

**Review by rasom — APPROVED** [review]

**Comment by churik** [comment]
@krisc 
#### 1) Part of header is not white (Android)
dsfdfsdfdsfdf
#### 2) After opening URL in native browser can't scan valid URL
Steps: 
- tap on + 
- scan QR code that contains link
- tap on open on IOS (or Android)
- go back to Status
- try to scan same QR with link
Expected result: can scan QR
Actual result: QR is not recognizable again

#### 3) Several times I saw crash on Android after scanning valid QR
Unfortunately didn't notice any specific steps to reproduce, during testing I scanned ~30 times different URLs and three times I saw this error anter tapping on "Open to Status" when scanning QR with link. Logcat: 
[crashlog.log](https://github.com/status-im/status-react/files/3084277/crashlog.log)

photo_2019-04-16 12 12 07

#### 4) Ethereum address is considered as a valid link.
I didn't expect that wallet address can be scanned with "Scan QR code" in chat.
In this case I would prefer to see error.
Example of QR:
photo_2019-04-16 12 14 49

**Comment by krisc** [comment]
@churik thanks for uncovering these issues. I'll have more time to look into them next week.

**Comment by krisc** [comment]
Sorry, I've been busy with Ethereal Hackathon these past few weeks. I'll have more time to get back on this after the weekend.

But these issues may be related to some code I wrote for this issue https://github.com/status-im/status-react/pull/7745#issuecomment-487472440

**Comment by flexsurfer** [comment]
hey @krisc any updates on this? :)

**Comment by krisc** [comment]
> hey @krisc any updates on this? :)

I was gonna try to resolve issues with https://github.com/status-im/status-react/pull/7745#issuecomment-487472440 because it might be related to the issues here.

Sorry I'm taking so long. I've been have some issues building and running since we switched to nix. So testing has been really slow for me.

**Comment by krisc** [comment]
I think I got my nix issues taken care of, but the easiest thing for me to move forward with this issue is to start a new PR. hope that's okay.

**Comment by flexsurfer** [comment]
yeah that's fine, thanks!

**Pull request closed without merging**